File: src/main/java/net/maple/packets/FieldPackets.java
package net.maple.packets;

import client.Character;
import field.object.FieldObject;
import field.object.drop.AbstractFieldDrop;
import field.object.drop.EnterType;
import net.maple.SendOpcode;
import util.packet.Packet;
import util.packet.PacketWriter;

public class FieldPackets {

    public static Packet setField(Character chr) {
        boolean isInstantiated = false;
        PacketWriter pw = new PacketWriter(32); // mhm hard to know

        pw.writeHeader(SendOpcode.SET_FIELD);
        pw.writeShort(0);
        pw.writeInt(chr.getChannel().getChannelId());
        pw.writeInt(0); // world

        pw.writeBool(true);
        pw.writeBool(!isInstantiated); // instantiated
        pw.writeShort(0);

        if (!isInstantiated) {
            pw.writeInt(0);
            pw.writeInt(0);
            pw.writeInt(0);

            CharacterPackets.encodeData(chr, pw);

            pw.writeInt(0);
            pw.writeInt(0);
            pw.writeInt(0);
            pw.writeInt(0);
        } else {
            System.err.println("[SetField] uuuh?");
        }

        pw.writeLong(System.currentTimeMillis() * 10000 + 116444592000000000L);

        return pw.createPacket();
    }

    public static Packet enterField(Character chr) {
        PacketWriter pw = new PacketWriter(32);

        pw.writeHeader(SendOpcode.USER_ENTER_FIELD);
        pw.writeInt(chr.getId()); // obj id

        pw.write(chr.getLevel());
        pw.writeMapleString(chr.getName());

        // guild
        pw.writeMapleString("");
        pw.writeShort(0);
        pw.write(0);
        pw.writeShort(0);
        pw.write(0);

        // temp stats
        // masks
        pw.writeInt(0);
        pw.writeInt(0);
        pw.writeInt(0);
        pw.writeInt(0);
        // nDefenseAtt & nDefenseState
        pw.write(0);
        pw.write(0);

        pw.writeShort(chr.getJob());

        CharacterPackets.encodeLooks(pw, chr, false);

        pw.writeInt(0);
        pw.writeInt(0);
        pw.writeInt(0);
        pw.writeInt(0);
        pw.writeInt(0); // complete set itemid
        pw.writeInt(chr.getPortableChair() == null ? 0 : chr.getPortableChair());

        pw.writePosition(chr.getPosition());
        pw.write(chr.getMoveAction());
        pw.writeShort(chr.getFoothold());
        pw.write(0); // ?

        // pets here

        // a whole bunch of ?
        pw.writeBool(false);
        pw.writeInt(0);
        pw.writeInt(0);
        pw.writeInt(0);
        pw.write(0);
        pw.writeBool(false);
        pw.writeBool(false);
        pw.writeBool(false);
        pw.writeBool(false);
        pw.write(0);
        pw.write(0);
        pw.writeInt(0);

        return pw.createPacket();
    }

    public static Packet leaveField(Character chr) {
        PacketWriter pw = new PacketWriter(6);

        pw.writeHeader(SendOpcode.USER_LEAVE_FIELD);
        pw.writeInt(chr.getId());

        return pw.createPacket();
    }

    public static Packet enterField(AbstractFieldDrop drop, byte enterType) {
        PacketWriter pw = new PacketWriter(32);
        //byte type = drop.getEnterType();
        FieldObject source = drop.getSource();

        pw.writeHeader(SendOpcode.DROP_ENTER_FIELD);
        pw.write(enterType);
        pw.writeInt(drop.getId());
        pw.writeBool(drop.isMeso());
        pw.writeInt(drop.getInfo());
        pw.writeInt(/*drop.getOwner()*/0);
        pw.write(/*type*/0x02); // own type
        pw.writePosition(drop.getPosition());
        pw.writeInt(source instanceof Character ? 0 : source.getId()); // source

        if (enterType != EnterType.FFA) {
            pw.writePosition(source.getPosition());
            pw.writeShort(0); // delay
        }

        if (!drop.isMeso()) {
            pw.writeLong(/*drop.getExpire() * 10000 + 116444592000000000L*/0);
        }

        pw.writeBool(false);
        pw.writeBool(false);

        return pw.createPacket();
    }

    public static Packet leaveField(AbstractFieldDrop drop) {
        return leaveField(drop, null);
    }

    public static Packet leaveField(AbstractFieldDrop drop, FieldObject source) {
        PacketWriter pw = new PacketWriter(14);

        pw.writeHeader(SendOpcode.DROP_LEAVE_FIELD);
        pw.write(drop.getLeaveType()); // nLeaveType
        pw.writeInt(drop.getId());

        if (drop.getLeaveType() == 0x02 || drop.getLeaveType() == 0x03 || drop.getLeaveType() == 0x05) {
            pw.writeInt(source == null ? 0 : source.getId());
        } else if (drop.getLeaveType() == 0x04) {
            pw.writeShort(0);
        }

        return pw.createPacket();
    }
}

File: src/main/java/net/maple/packets/FieldPackets.kt
package net.maple.packets

import client.Avatar
import client.Character
import client.effects.FieldEffectInterface
import field.Field
import field.obj.FieldObject
import field.obj.drop.AbstractFieldDrop
import field.obj.drop.EnterType
import net.maple.SendOpcode
import net.maple.packets.CharacterPackets.encodeData
import net.maple.packets.CharacterPackets.encodeLooks
import util.packet.Packet
import util.packet.PacketWriter
import kotlin.math.ceil

object FieldPackets {

    fun Character.setField(): Packet {
        val pw = PacketWriter(32)

        pw.writeHeader(SendOpcode.SET_FIELD)
        pw.writeShort(0)
        pw.writeInt(this.getChannel().channelId)
        pw.writeInt(0) // world

        fieldKey = ceil(Math.random() * 255).toInt().toByte()
        pw.writeByte(fieldKey)
        pw.writeBool(migrating)
        pw.writeShort(0) // chatblock?

        if (migrating) {
            migrating = false
            pw.writeInt(0) // calc seed
            pw.writeInt(0) // calc seed
            pw.writeInt(0) // chatblock?

            encodeData(pw)

            pw.writeInt(0)
            pw.writeInt(0)
            pw.writeInt(0)
            pw.writeInt(0)
        } else {
            pw.writeByte(0)
            pw.writeInt(fieldId)
            pw.writeByte(portal)
            pw.writeInt(health)

            pw.writeBool(chasing)
            if (chasing) {
                chasing = false
                pw.writeInt(position.x)
                pw.writeInt(position.y)
            }
        }

        pw.writeLong(System.currentTimeMillis() * 10000 + 116444592000000000L)
        return pw.createPacket()
    }

    fun Avatar.enterField(): Packet {
        val pw = PacketWriter(32)

        pw.writeHeader(SendOpcode.USER_ENTER_FIELD)
        pw.writeInt(this.id) // obj id
        pw.write(this.level)
        pw.writeMapleString(this.name)

        // guild
        this.guild?.let {
            pw.writeMapleString(it.name)
            it.mark?.encode(pw) ?: pw.write(ByteArray(6))
        } ?: pw.writeMapleString(if (this is Character) "" else "Replay").write(ByteArray(6))

        // temp stats
        // masks
        pw.writeInt(0)
        pw.writeInt(0)
        pw.writeInt(0)
        pw.writeInt(0)

        // nDefenseAtt & nDefenseState
        pw.write(0)
        pw.write(0)

        pw.writeShort(this.job)

        encodeLooks(pw, false)

        pw.writeInt(0)
        pw.writeInt(0)
        pw.writeInt(0)
        pw.writeInt(0)
        pw.writeInt(0) // complete set itemid

        pw.writeInt(this.portableChair ?: 0)

        pw.writePosition(this.position)
        pw.write(this.moveAction.toInt())
        pw.writeShort(this.foothold)
        pw.write(0) // ?

        pets.forEach {
            pw.writeBool(true)
            it.encodeData(pw)
        }

        // a whole bunch of ?
        pw.writeBool(false)
        pw.writeInt(0)
        pw.writeInt(0)
        pw.writeInt(0)
        pw.write(0)
        pw.writeBool(false)
        pw.writeBool(false)
        pw.writeBool(false)
        pw.writeBool(false)
        pw.write(0)
        pw.write(0)
        pw.writeInt(0)

        return pw.createPacket()
    }

    fun Avatar.leaveField(): Packet {
        val pw = PacketWriter(6)

        pw.writeHeader(SendOpcode.USER_LEAVE_FIELD)
        pw.writeInt(this.id)

        return pw.createPacket()
    }

    fun AbstractFieldDrop.enterField(enterType: Byte, cursedObject: Int? = null): Packet {
        val pw = PacketWriter(32)

        //byte type = drop.getEnterType();
        val source = this.source

        pw.writeHeader(SendOpcode.DROP_ENTER_FIELD)
        pw.write(enterType.toInt())
        pw.writeInt(this.id)
        pw.writeBool(this.isMeso)
        pw.writeInt(cursedObject ?: this.info)
        pw.writeInt( /*drop.getOwner()*/0)
        pw.write( /*type*/0x02) // own type
        pw.writePosition(this.position)
        pw.writeInt(if (source is Character) 0 else source.id) // source

        if (enterType != EnterType.FFA) {
            pw.writePosition(source.position)
            pw.writeShort(0) // delay
        }

        if (!this.isMeso) {
            pw.writeLong( /*drop.getExpire() * 10000 + 116444592000000000L*/0)
        }

        pw.writeBool(false)
        pw.writeBool(false)

        return pw.createPacket()
    }

    fun AbstractFieldDrop.leaveField(source: FieldObject? = null): Packet {
        val pw = PacketWriter(14)

        pw.writeHeader(SendOpcode.DROP_LEAVE_FIELD)
        pw.write(this.leaveType.toInt()) // nLeaveType
        pw.writeInt(this.id)

        if (this.leaveType.toInt() == 0x02 || this.leaveType.toInt() == 0x03 || this.leaveType.toInt() == 0x05) {
            pw.writeInt(source?.id ?: 0)
        } else if (this.leaveType.toInt() == 0x04) {
            pw.writeShort(0)
        }

        return pw.createPacket()
    }

    fun Field.fieldEffect(effect: FieldEffectInterface) {
        val pw = PacketWriter(12)

        pw.writeHeader(SendOpcode.FIELD_EFFECT)
        effect.encode(pw)

        this.broadcast(pw.createPacket())
    }
}